Boost: Update LCP Error Messaging within Plugin

LiamSarsfield opened the pull request:
Fixes [HOG-125: Make LCP Errors Surfaced from Cloud more Readable](https://linear.app/a8c/issue/HOG-125/make-lcp-errors-surfaced-from-cloud-more-readable)

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Uses specific types which are determined by the Cloud to display what error messages the user will see, instead of determining the error message string directly from the Cloud

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
**Prerequisites:**

- Ensure https://github.com/Automattic/boost-cloud/pull/691 is checked out within Boost Cloud
- Latest Jetpack Boost Developer plugin installed and activated.
- Your local Boost Cloud is connected to your local WP instance.
- At least one Cornerstone Page set.
- Boost Feature Flag enabled within the install (`JETPACK_BOOST_DEVELOPMENT_FEATURES`)
- LCP Optimization feature enabled.

**HTTP Error:**
1. Within the Boost Developer, under "Force Critical CSS generation Error", set this to an option such as "Advanced Recommendations".
2. Run the LCP Optimization, and wait for it to be completed.
3. You should see at least one issue (or multiple if multiple Cornerstone pages) appear underneath LCP Optimization per below:
<img width="753" alt="Screenshot 2025-06-03 at 12 02 42" src="https://github.com/user-attachments/assets/161a56b9-96ad-4115-a273-056a92c85060" />

**Timeout Error:**
1. Ensure the "Force Critical CSS generation Error" within the Boost developer Settings is set up "None"
2. Attach the below HTML snippet to your WP site.
  <details>
  <summary>HTML</summary>

  ```php
		add_filter(
			'the_content',
			function (
				$content
			) {
				// If the ?jb-disable-modules=lcp parameter is present, sleep for 61 seconds.
				if ( ! empty( $_GET['jb-disable-modules'] ) ) {
					sleep( 61 );
				}
				return $content;
			}
		);
  ```

  </details>

3. Run the LCP Optimization, and wait for it to be completed.
4. You should see at least one issue (or multiple if multiple Cornerstone pages) appear underneath LCP Optimization per below:
<img width="891" alt="Screenshot 2025-06-03 at 12 40 50" src="https://github.com/user-attachments/assets/17e910ab-9f62-4d7e-8ec2-4e33d9ba431f" />

LiamSarsfield committed:
Boost: Enhance LCP error handling with detailed error types and messages

- Introduced new LcpErrorTypeSchema to categorize LCP errors.
- Updated ErrorDetails component to display user-friendly error messages based on error types.
- Modified LCP data structure to replace 'message' with 'type' for error reporting.
- Adjusted API response handling to accommodate new error structure.

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/lcp/error-displaying

LiamSarsfield committed:
Refactor ErrorDetails component to improve error message handling

- Updated error message formatting for clarity.
- Changed variable name from `errors` to `pagesWithErrors` for better readability.

LiamSarsfield committed:
Add 'lcp-metric-timeout' error type and corresponding message to ErrorDetails component

- Introduced a new error type 'lcp-metric-timeout' to LcpErrorTypeSchema.
- Updated ErrorDetails component to include a user-friendly message for the new error type, enhancing error handling and user experience.

LiamSarsfield committed:
Enhance LCP feature description and title for clarity in user experience (#43745)

* Updated the title to "Optimize LCP Images" for better specificity.
* Revised the description to emphasize the importance of optimizing key images for faster loading of main content on Cornerstone pages.

LiamSarsfield committed:
changelog

kraftbj commented on projects/plugins/boost/app/assets/src/js/features/lcp/status/error-details.tsx:
```diff
@@ -6,6 +6,29 @@ import { useLcpState } from '../lib/stores/lcp-state';
+			'jetpack-boost'
+		),
+		'element-not-unique': __(
+			'This page has multiple similar large elements, making it difficult to determine which one to optimize. Manual optimization may be needed.',
```
If I was an user, what would "manual optimization" mean? Do we have any documentation to explain that which we could surface here? Could we reword it to state a potential action, etc?

LiamSarsfield replied:
This error message was introduced in https://github.com/Automattic/boost-cloud/pull/680, but it should be extremely unlikely to occur after https://github.com/Automattic/boost-cloud/pull/685 was merged.
I've updated it (and all error types) to include a "Learn more" button which redirects to the LCP documentation with further information on what fixing this may entail:
https://jetpack.com/support/jetpack-boost/optimize-your-largest-contentful-paint-lcp-with-jetpack-boost/#element-not-unique

I agree though that the "manual optimization" is too broad and may confuse the user about a simplistic solution, when realistically their page is likely too complex to address it properly, and maintain their page's complex functionality.
<img width="728" alt="Screenshot 2025-06-04 at 13 01 29" src="https://github.com/user-attachments/assets/288ee6d0-9bde-4545-b2e7-1e815592ada0" />

LiamSarsfield committed:
Refine LCP feature description for improved clarity and user engagement. Updated text to emphasize the optimization of key images for a smoother experience on Cornerstone Pages.

LiamSarsfield committed:
Merge branch 'update/boost/lcp/description' into update/boost/lcp/error-displaying

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/status/error-details.tsx:
```diff
@@ -6,6 +6,29 @@ import { useLcpState } from '../lib/stores/lcp-state';
+			'jetpack-boost'
+		),
+		'lcp-timeout': __(
+			'The page took too long to load during optimization. Please check that the page is publicly accessible and try again.',
```
Nitpick: Technically, the page failed to be analyzed. The optimization is the final step.

LiamSarsfield replied:
Updated 👍

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/status/error-details.tsx:
```diff
@@ -6,6 +6,29 @@ import { useLcpState } from '../lib/stores/lcp-state';
 export const ErrorDetails = () => {
+	const dictionary = {
+		unknown: __(
+			'Something went wrong while optimizing this page. Please try again later, or contact support if the issue persists.',
```
I really hope this doesn't end up being "failed to generate critical css" for LCP 😂

dilirity reviewed (changes requested):
Nice work! I think overall this is better than before.

Once thing I noticed is that throughout Boost we word our messages in various forms. For example:
- if something cache related fails we use passive voice - `Failed to write settings file ...`
- sometimes we use third person (branding) - `Jetpack Boost couldn't verify this page:`
- the current messages use first-person plural ("we") - "We something"

I think things will read better if we aim for a consistent voice and tone.

For now, I think it makes sense to update these messages only (if what I'm saying actually makes sense, since I'm not a native speaker 😅). For the rest, we can update as we go.

dilirity commented:
Something that occurred to me after the review - since we have error codes, we could tie those in with the documentation (like we have for Cloud CSS and Page Cache errors). If it makes sense that is.

For example, page cache troubleshooting page [here](https://jetpack.com/support/jetpack-boost/troubleshooting-caching-issues/). Notice the sections on the side - the anchors are the same as the error codes Boost returns.

dilirity reviewed (commented):
I noticed that the cloud response returns meta that has either code or selector, but they aren't used here.

Can they be included to give our users more info they can provide if they need help?

LiamSarsfield committed:
Enhance error messages in ErrorDetails component for clarity and user guidance. Updated text to provide more specific context regarding connection issues and loading times, and added a new link to documentation for further assistance.

LiamSarsfield committed:
Add event tracking for external link in ErrorDetails component

- Implemented a click event handler to record user interactions with the 'Learn More' link, enhancing analytics for error type engagement.

LiamSarsfield commented:
@dilirity 
> Once thing I noticed is that  throughout Boost we word our messages in various forms.

Great points, this makes complete sense, and I believe this is related to having a consistent [Information Architecture](https://www.uxdesigninstitute.com/blog/what-is-information-architecture/) UX.

I've removed the "We"/ first-person plural phrasing, and replaced it with "Jetpack Boost"/third person phrasing.
It's something we definitely should watch out for, thanks for spotting.

LiamSarsfield commented:
> Something that occurred to me after the review - since we have error codes, we could tie those in with the documentation (like we have for Cloud CSS and Page Cache errors). If it makes sense that is.

I've applied this now. I've created a private page for the LCP feature below, and created the appropriate anchor points that refer to each type:
https://jetpack.com/support/jetpack-boost/optimize-your-largest-contentful-paint-lcp-with-jetpack-boost/#specific-error-types

Each bullet point/Error message will include a "Learn more" button which links to their appropriate error.

LiamSarsfield committed:
Merge branch 'trunk' into update/boost/lcp/error-displaying

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/status/error-details.tsx:
```diff
@@ -18,12 +44,36 @@ export const ErrorDetails = () => {
-	const errorMessages = errors.flatMap( p => ( p.errors || [] ).map( e => e.message ) );
+	const errorMessages: ReactElement[] = [];
+
+	pagesWithErrors.forEach( page => {
+		page.errors?.forEach( error => {
+			if ( error?.type ) {
+				errorMessages.push(
+					createInterpolateElement(
+						`${ dictionary[ error?.type ] } (${ page.url }) <link>Learn more</link>`,
+						{
+							link: (
+								<ExternalLink
+									href={ `https://jetpack.com/support/jetpack-boost/optimize-your-largest-contentful-paint-lcp-with-jetpack-boost/#${ error?.type }` }
+									onClick={ () => {
+										recordBoostEvent( 'lcp_learn_more', {
+											error_type: error?.type,
+										} );
+									} }
+								/>
+							),
+						}
+					)
+				);
+			}
+		} );
+	} );
```
I feel like the nesting here is a bit excessive 😅 

Since we're not translating the string, do we need `createInterpolateElement`?

Also - make sure to always use the Jetpack redirect service. I updated the URL for the page earlier and now it returns a 404 when visiting any error code:

https://jetpack.com/support/jetpack-boost/optimize-your-largest-contentful-paint-lcp-with-jetpack-boost/#http-error

dilirity replied:
I wonder if `reduce` would help along with an early return or something 🤔

LiamSarsfield replied:
It does seem pretty excessive on GitHub alright. I've updated it now to not use nested forEachs, and also to use the Redirect service.
In a separate PR we should also update C.CSS' link as well
https://github.com/Automattic/jetpack/blob/update/boost/lcp/error-displaying/projects/plugins/boost/app/assets/src/js/lib/utils/get-support-link-critical-css.ts#L20-L24

dilirity reviewed (changes requested):
A couple of things and we should be good to go!

kraftbj reviewed (changes requested):
I agree with Pete. Regarding the `unknown`, let's be sure to have a tracking plan to monitor the types of errors happening so if we see a lot of unknowns, we can have a swing at identifying some of them.

LiamSarsfield committed:
Enhance LCP error handling by introducing LcpErrorMetaSchema and updating ErrorDetails component

- Added LcpErrorMetaSchema to define optional metadata for error types, including code and selector.
- Refactored ErrorDetails component to utilize the new schema, improving clarity and user guidance for various error scenarios.

LiamSarsfield committed:
Merge branch 'update/boost/lcp/error-displaying' of github.com:Automattic/jetpack into update/boost/lcp/error-displaying

LiamSarsfield committed:
Refactor error link in ErrorDetails component to use getRedirectUrl function

- Updated the external link in the PageError component to utilize the getRedirectUrl function for improved URL management.

LiamSarsfield commented:
@kraftbj 

> let's be sure to have a tracking plan to monitor the types of errors happening so if we see a lot of unknowns, we can have a swing at identifying some of them.

Would you have a suggestion on a suitable way to do this? I'm not familiar with an existing pattern we may have for this.
Currently the "Learn more" button is being tracked (including unknown error types), so that may give us an indication of the current trend of common errors users are experiencing, providing the user clicks the link, that is.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/status/error-details.tsx:
```diff
@@ -1,10 +1,86 @@
+			);
+		}
+
+		if ( type === 'element-not-unique' && meta?.selector !== undefined ) {
+			return sprintf(
+				/* translators: %s is the HTML selector */
+				__(
+					'This page has multiple similar large elements with the same selector (%s), making it difficult to determine which one to optimize.',
+					'jetpack-boost'
+				),
+				meta.selector
+			);
+		}
```
Updated check and messaging.

```suggestion
		if ( type === 'element-not-unique' ) {
			return sprintf(
				/* translators: %s is the HTML selector */
				__(
					'Boost was unable to identify the main LCP element on the page, due to multiple similar large elements with the same selector (%s).',
					'jetpack-boost'
				),
				meta?.selector ?? 'unknown'
			);
		}
```

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/status/error-details.tsx:
```diff
@@ -1,10 +1,86 @@
+
+const PageError = ( { url, error }: PageErrorProps ) => {
+	const getErrorLabel = ( { type, meta }: LcpErrorDetails ) => {
+		if ( type === 'http-error' && meta?.code !== undefined ) {
+			return createInterpolateElement(
+				sprintf(
+					/* translators: %d is the HTTP Status Code */
+					__(
+						'Boost received HTTP error <b>%d</b> when attempting to analyze this page. Please check that the page is publicly accessible, and try again.',
+						'jetpack-boost'
+					),
+					meta.code
+				),
+				{
+					b: <strong />,
+				}
+			);
+		}
```
Let's only check the type here.

The reason for this being - if something happens and there's no `meta.code`, an incomplete http error would show the generic one.

But if it's an http error with an unknown code, that at least gives us some idea that it's at least related to the loading of the page.

I've also updated the messaging.

```suggestion
		if ( type === 'http-error' ) {
			return createInterpolateElement(
				sprintf(
					/* translators: %d is the HTTP Status Code */
					__(
						'Boost encountered an HTTP error <b>%d</b> when attempting to analyze this page. Please make sure it loads correctly in incognito mode, and try again.',
						'jetpack-boost'
					),
					meta?.code ?? 'unknown'
				),
				{
					b: <strong />,
				}
			);
		}
```

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/status/error-details.tsx:
```diff
@@ -1,10 +1,86 @@
+			);
+		}
+
+		if ( type === 'lcp-timeout' ) {
+			return __(
+				'The page took too long to load during analysis. Please check that the page is publicly accessible, and try again.',
+				'jetpack-boost'
+			);
+		}
```
Updated messaging.

```suggestion
		if ( type === 'lcp-timeout' ) {
			return __(
				'The page took too long to load during analysis. Please make sure the page loads correctly in incognito mode, and try again.',
				'jetpack-boost'
			);
		}
```

dilirity reviewed (changes requested):
Thanks for addressing the feedback!

A few small notes and we can merge this.

LiamSarsfield committed:
Update LCP 'element-not-unique' messaging

Co-authored-by: Peter Petrov <peter.petrov89@gmail.com>

LiamSarsfield committed:
Update LCP 'lcp-timeout' messaging

Co-authored-by: Peter Petrov <peter.petrov89@gmail.com>

LiamSarsfield committed:
Update LCP ''http-error'' messaging

Co-authored-by: Peter Petrov <peter.petrov89@gmail.com>

dilirity reviewed (approved):
:shipit:
